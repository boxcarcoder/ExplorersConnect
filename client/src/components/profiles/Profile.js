import React, { Fragment, useEffect } from 'react';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getProfileById } from '../../actions/profile';

import Spinner from '../layout/Spinner';

const Profile = ({
  profileState: { profile, loading },
  getProfileById,
  match,
}) => {
  //when a user's profile page is first loaded, retrieve the profile and store it in the profile redux state
  useEffect(() => {
    getProfileById(match.params.id);
  }, [getProfileById, match.params.id]);

  //for async operation. need to wait for profile to be loaded into the profile redux state before proceeding
  if (profile === null || loading) {
    return (
      <Fragment>
        <Spinner />
      </Fragment>
    );
  }

  const {
    user: { name, avatar }, //this field countains a user field, which contains name and avatar
    bio,
    location,
    Hiking,
    Camping,
    Kayaking,
    Rafting,
    Skiing,
    Snowboarding,
    Rockclimbing,
    destinations,
    gears,
    social,
  } = profile;

  // ======== Check for favorite activities ========
  const checkHike = () => {
    if (Hiking) {
      return (
        <div className='p-1'>
          <i className='fas fa-campground'></i> Camping
        </div>
      );
    }
  };

  const checkCamp = () => {
    if (Camping) {
      return (
        <div className='p-1'>
          <i className='fas fa-hiking'></i> Hiking
        </div>
      );
    }
  };

  const checkKayak = () => {
    if (Kayaking) {
      return (
        <div className='p-1'>
          <i className='fas fa-water'></i> Kayaking
        </div>
      );
    }
  };

  const checkRaft = () => {
    if (Rafting) {
      return (
        <div className='p-1'>
          <i className='fas fa-tint'></i> Rafting
        </div>
      );
    }
  };

  const checkSki = () => {
    if (Skiing) {
      return (
        <div className='p-1'>
          <i className='fas fa-skiing'></i> Skiing
        </div>
      );
    }
  };

  const checkSnowboard = () => {
    if (Snowboarding) {
      return (
        <div className='p-1'>
          <i className='fas fa-snowboarding'></i> Snowboarding
        </div>
      );
    }
  };

  const checkRockclimb = () => {
    if (Rockclimbing) {
      return (
        <div className='p-1'>
          <i className='fas fa-mountain'></i> Rockclimbing
        </div>
      );
    }
  };

  // ======== Check for favorite destinations for each activity ========
  const checkHikeDestinations = () => {
    if (Hiking) {
      return (
        <Fragment>
          <p>
            <strong>Hike Trails</strong>
            <ul>
              {destinations.map((destination) => (
                <li>{destination.hikingTrails}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkCampDestinations = () => {
    if (Camping) {
      return (
        <Fragment>
          <p>
            <strong>Campsites</strong>
            <ul>
              {destinations.map((destination) => (
                <li>{destination.campSites}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkWaterDestinations = () => {
    if (Kayaking || Rafting) {
      return (
        <Fragment>
          <p>
            <strong>Lakes, Rivers, or Oceans</strong>
            <ul>
              {destinations.map((destination) => (
                <li>{destination.waterAreas}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkSnowDestinations = () => {
    if (Skiing || Snowboarding) {
      return (
        <Fragment>
          <p>
            <strong>Slopes</strong>
            <ul>
              {destinations.map((destination) => (
                <li>{destination.slopes}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkClimbDestinations = () => {
    if (Rockclimbing) {
      return (
        <Fragment>
          <p>
            <strong>Crags</strong>
            <ul>
              {destinations.map((destination) => (
                <li>{destination.crags}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  // ======== Check for favorite gears for each activity ========
  const checkHikeGears = () => {
    if (Hiking) {
      return (
        <Fragment>
          <p>
            <strong>Hiking</strong>
            <ul>
              {gears.map((gear) => (
                <li>{gear.hikeGear}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkCampGears = () => {
    if (Camping) {
      return (
        <Fragment>
          <p>
            <strong>Camping</strong>
            <ul>
              {gears.map((gear) => (
                <li>{gear.campGear}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkWaterGears = () => {
    if (Kayaking || Rafting) {
      return (
        <Fragment>
          <p>
            <strong>Lakes, Rivers, or Oceans</strong>
            <ul>
              {gears.map((gear) => (
                <li>{gear.waterGear}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkSnowGears = () => {
    if (Skiing || Snowboarding) {
      return (
        <Fragment>
          <p>
            <strong>Slopes</strong>
            <ul>
              {gears.map((gear) => (
                <li>{gear.snowGear}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  const checkClimbGears = () => {
    if (Rockclimbing) {
      return (
        <Fragment>
          <p>
            <strong>Rock Climbing</strong>
            <ul>
              {gears.map((gear) => (
                <li>{gear.rockClimbingGear}</li>
              ))}
            </ul>
          </p>
          <div className='line'></div>
        </Fragment>
      );
    }
  };

  // ======== Check for provided social media links ========
  const checkInstagram = () => {
    if (!social || !social.instagram) {
      return (
        <div className='invalid-icon'>
          <i className='fab fa-instagram fa-2x'></i>
        </div>
      );
    } else if (social.instagram) {
      return (
        <div className='valid-icon'>
          <a href={social.instagram} target='_blank' rel='noopener noreferrer'>
            <i className='fab fa-instagram fa-2x'></i>
          </a>
        </div>
      );
    }
  };

  const checkTwitter = () => {
    if (!social || !social.twitter) {
      return (
        <div className='invalid-icon'>
          <i className='fab fa-twitter fa-2x'></i>
        </div>
      );
    } else if (social.twitter) {
      return (
        <div className='valid-icon'>
          <a href={social.twitter} target='_blank' rel='noopener noreferrer'>
            <i className='fab fa-twitter fa-2x'></i>
          </a>
        </div>
      );
    }
  };

  const checkFacebook = () => {
    if (!social || !social.facebook) {
      return (
        <div className='invalid-icon'>
          <i className='fab fa-facebook fa-2x'></i>
        </div>
      );
    } else if (social.facebook) {
      return (
        <div className='valid-icon'>
          <a href={social.facebook} target='_blank' rel='noopener noreferrer'>
            <i className='fab fa-facebook fa-2x'></i>
          </a>
        </div>
      );
    }
  };

  const checkYoutube = () => {
    if (!social || !social.youtube) {
      return (
        <div className='invalid-icon'>
          <i className='fab fa-youtube fa-2x'></i>
        </div>
      );
    } else if (social.youtube) {
      return (
        <div className='valid-icon'>
          <a href={social.youtube} target='_blank' rel='noopener noreferrer'>
            <i className='fab fa-youtube fa-2x'></i>
          </a>
        </div>
      );
    }
  };

  const checkSocialMedia = () => {
    return (
      <Fragment>
        <div className='vert-m-1 icons-list'>
          {checkInstagram()}
          {checkTwitter()}
          {checkFacebook()}
          {checkYoutube()}
        </div>
      </Fragment>
    );
  };

  return (
    <Fragment>
      <Link to='/profiles' className='btn'>
        Back to Profiles
      </Link>
      <div className='vert-m-1 profile-grid'>
        {/* Top */}
        <div className='bg-primary p-2 profile-top'>
          <img className='curved-img vert-m-1' src={avatar} alt='profilePic' />

          <h1 className='large'>{name}</h1>
          <p>{location}</p>
          {checkSocialMedia()}
        </div>

        {/* Bio and Passions */}
        <div className='bg-light p-2 profile-about'>
          <h2 className='text-primary'>Bio</h2>
          <p>{bio}</p>

          <div className='line'></div>
          <h2 className='text-primary'>Passions</h2>
          <div className='passions'>
            {checkHike()}
            {checkCamp()}
            {checkKayak()}
            {checkRaft()}
            {checkSki()}
            {checkSnowboard()}
            {checkRockclimb()}
          </div>
        </div>

        {/* Destinations */}
        <div className='bg-light p-2 profile-location'>
          <h2 className='text-primary'>Explored</h2>
          {checkHikeDestinations()}
          {checkCampDestinations()}
          {checkWaterDestinations()}
          {checkSnowDestinations()}
          {checkClimbDestinations()}
        </div>

        {/* Gears */}
        <div className='bg-light p-2 profile-gear'>
          <h2 className='text-primary'>Gear</h2>
          {checkHikeGears()}
          {checkCampGears()}
          {checkWaterGears()}
          {checkSnowGears()}
          {checkClimbGears()}
        </div>
      </div>
    </Fragment>
  );
};

Profile.propTypes = {
  profileState: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  profileState: state.profile,
});

export default connect(mapStateToProps, { getProfileById })(Profile);
