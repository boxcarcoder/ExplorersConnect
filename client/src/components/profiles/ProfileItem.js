import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

import PropTypes from 'prop-types';

const ProfileItem = ({
  profile: {
    user: { _id, name, avatar },
    Hiking,
    Camping,
    Kayaking,
    Rafting,
    Skiing,
    Snowboarding,
    Rockclimbing,
    Other,
    location,
    bio
  }
}) => {

  const checkHike = () => {
    if (Hiking) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Hiking
        </li>
      );
    }
  };

  const checkCamp = () => {
    if (Camping) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Camping
        </li>
      );
    }
  };

  const checkKayak = () => {
    if (Kayaking) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Kayaking
        </li>
      );
    }
  };

  const checkRaft = () => {
    if (Rafting) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Rafting
        </li>
      );
    }
  };

  const checkSki = () => {
    if (Skiing) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Skiing
        </li>
      );
    }
  };

  const checkSnowboard = () => {
    if (Snowboarding) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Snowboarding
        </li>
      );
    }
  };

  const checkRockclimb = () => {
    if (Rockclimbing) {
      return (
        <li className='text-primary'>
          <i className='fas fa-check'></i> Rockclimbing
        </li>
      );
    }
  };

  return (
    <Fragment>
      <div className='profile bg-light'>
        <img className='round-img' src={avatar} alt='hiker' />
        <div className='m-3'>
          <h2>{name}</h2>
          <p>{location}</p>
          <p>{bio}</p>
          <Link to='/profile' className='btn btn-primary'>
            View Profile
          </Link>
        </div>
        <ul>
          {checkHike()}
          {checkCamp()}
          {checkKayak()}
          {checkRaft()}
          {checkSki()}
          {checkSnowboard()}
          {checkRockclimb()}
        </ul>
      </div>
    </Fragment>
  );
};

ProfileItem.propType = {
  profile: PropTypes.object.isRequired
};

export default ProfileItem;
