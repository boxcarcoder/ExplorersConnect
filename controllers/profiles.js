const profilesRouter = require('express').Router();
const mongoose = require('mongoose');
const Profile = require('../models/Profile');
const auth = require('../middlewares/auth');
const { check, validationResult } = require('express-validator');

// @route   GET api/profiles/me
// @desc    Fetch a user's profile
// @access  Private
profilesRouter.get('/me', auth, async (req, res) => {
  try {
    // find the profile of a requested user
    let profile = await Profile.findOne({ user: req.user.id });

    if (!profile) {
      res.status(400).json('This profile does not exist.');
    }

    res.json(profile);
  } catch (err) {
    console.error(err.message);
    res.status(500).json('Server error.');
  }
});

// @route   POST api/profiles
// @desc    Create a profile for a user
// @access  Private
profilesRouter.post(
  '/',
  [
    auth,
    [
      check('bio', 'Please provide a bio')
        .not()
        .isEmpty(),
      check('country', 'Please provide your current country')
        .not()
        .isEmpty()
    ]
  ],
  async (req, res) => {
    let errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    // create a profile for a user
    const {
      bio,
      country,
      //passions,
      gears,
      youtube,
      twitter,
      facebook,
      instagram
    } = req.body;

    //Build object for the profile's fields
    let profileFields = {};
    profileFields.user = req.user.id;
    if (bio) profileFields.bio = bio;
    if (country) profileFields.country = country;
    if (gears) {
      profileFields.gears = gears.split(',').map(gear => gear.trim()); //turn comma separated string into array by split, then map to remove spaces from each index
    }

    //Build social object for our profile's fields object
    profileFields.social = {};
    if (youtube) profileFields.social.youtube = youtube;
    if (twitter) profileFields.social.twitter = twitter;
    if (facebook) profileFields.social.facebook = facebook;
    if (instagram) profileFields.social.instagram = instagram;

    try {
      let profile = await Profile.findOne({ user: req.user.id });

      if (profile) {
        //update the profile
        profile = await Profile.findOneAndUpdate(
          { user: req.user.id },
          { $set: profileFields },
          { new: true }
        );
        return res.json(profile);
      }

      // Create a new profile
      profile = new Profile(profileFields);
      await profile.save();
      res.json(profile);
    } catch (err) {
      console.error(err.message);
      res.status(500).json('Server error.');
    }
  }
);

module.exports = profilesRouter;

// if (passions) {
//   profileFields.passions = passions
//     .split(',')
//     .map(passion => passion.trim());
// }
