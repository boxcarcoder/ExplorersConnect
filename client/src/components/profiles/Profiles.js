import React, { Fragment } from 'react';
import ProfileItem from './ProfileItem';
import Spinner from '../layout/Spinner';

//redux
import { connect } from 'react-redux';
import { getAllProfiles } from '../../actions/profile';
import PropTypes from 'prop-types';

export const Profiles = ({
  getAllProfiles,
  profileState: { profiles, loading },
}) => {
  //on the Profiles page's first load, retrieve all profiles and save it into the profile redux state
  React.useEffect(() => {
    getAllProfiles();
  }, []); //*** */

  if (!profiles.length || loading) {
    return <Spinner />;
  }
  return (
    <Fragment>
      <h1 className='large text-primary'>Explorers</h1>
      <p className='lead'>
        <i className='fab fa-connectdevelop'></i> Browse and connect with
        explorers
      </p>
      <div className='profiles'>
        {profiles.map((profile) => (
          <ProfileItem key={profile._id} profile={profile} />
        ))}
      </div>
    </Fragment>
  );
};

Profiles.propTypes = {
  getAllProfiles: PropTypes.func.isRequired,
  profileState: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  profileState: state.profile,
});

export default connect(mapStateToProps, { getAllProfiles })(Profiles);
