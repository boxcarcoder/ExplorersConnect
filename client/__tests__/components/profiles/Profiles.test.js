import React from 'react';
import { shallow } from 'enzyme';

// Import the non-connected component
import { Profiles } from '../../../src/components/profiles/Profiles';

// Globals for testing
let wrapper;
let props;
let mockGetAllProfiles = jest.fn();


describe('<Profiles /> component.', () => {
    describe('Successfully renders', () => {
        test('the component.', () => {
            props = {
                profileState: {
                    profiles: [
                        {
                            user: 'test user id',
                            bio: 'test user bio',
                            location: 'test user location',
                        },
                        {
                            user: 'test user2 id',
                            bio: 'test user2 bio',
                            location: 'test user2 location',
                        },
                    ],
                    loading: false,
                },
                getAllProfiles: mockGetAllProfiles
            };
    
            wrapper = shallow(<Profiles {...props}/>);
            expect(wrapper).toMatchSnapshot();
        });

        test('a spinner if the loading property in the Profile state is true.', () => {
            // Set loading to true
            props = {
                profileState: {
                    profiles: [
                        {
                            user: 'test user id',
                            bio: 'test user bio',
                            location: 'test user location',
                        },
                        {
                            user: 'test user2 id',
                            bio: 'test user2 bio',
                            location: 'test user2 location',
                        },
                    ],
                    loading: true,
                },
                getAllProfiles: mockGetAllProfiles
            };
    
            wrapper = shallow(<Profiles {...props}/>);

             // Execute the test.
            const spinner = wrapper.find('Spinner');
            expect(spinner.exists()).toBe(true);
        });

        test('a spinner if there are no profiles in the Profiles state.', () => {
            // Set profiles to be empty.
            props = {
                profileState: {
                    profiles: [],
                    loading: false,
                },
                getAllProfiles: mockGetAllProfiles
            };
    
            wrapper = shallow(<Profiles {...props}/>);

            // Execute the test.
            const spinner = wrapper.find('Spinner');
            expect(spinner.exists()).toBe(true);
        });

        test('a ProfileItem component for each profile in the Profile state.', () => {
            props = {
                profileState: {
                    profiles: [
                        {
                            user: 'test user id',
                            bio: 'test user bio',
                            location: 'test user location',
                        },
                        {
                            user: 'test user2 id',
                            bio: 'test user2 bio',
                            location: 'test user2 location',
                        },
                    ],
                    loading: false,
                },
                getAllProfiles: mockGetAllProfiles
            };
    
            wrapper = shallow(<Profiles {...props}/>);

            // Execute the test.
            const profileItem1 = wrapper.find('ProfileItem').at(0);
            expect(profileItem1.exists()).toBe(true);

            const profileItem2 = wrapper.find('ProfileItem').at(1);
            expect(profileItem2.exists()).toBe(true);
        });
    });

    test('The getAllProfiles() action is fired after the first render.', () => {
        props = {
            profileState: {
                profiles: [
                    {
                        user: 'test user id',
                        bio: 'test user bio',
                        location: 'test user location',
                    },
                    {
                        user: 'test user2 id',
                        bio: 'test user2 bio',
                        location: 'test user2 location',
                    },
                ],
                loading: false,
            },
            getAllProfiles: mockGetAllProfiles
        };


        // Spy on the useEffect method.
        let useEffect = jest.spyOn(React, 'useEffect');

        // Re-define the useEffect method to run syncrhonously.
        let mockUseEffect = () => {
            useEffect.mockImplementationOnce((f) => f());
        }

        // Execute the test.
        mockUseEffect();
        wrapper = shallow(<Profiles {...props}/>);

        expect(mockGetAllProfiles).toHaveBeenCalled();
    });
});