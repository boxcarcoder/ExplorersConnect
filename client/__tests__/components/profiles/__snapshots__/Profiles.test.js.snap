// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profiles /> component. Successfully renders the component. 1`] = `
<Fragment>
  <h1
    className="large text-primary"
  >
    Explorers
  </h1>
  <p
    className="lead"
  >
    <i
      className="fab fa-connectdevelop"
    />
     Browse and connect with explorers
  </p>
  <div
    className="profiles"
  >
    <ProfileItem
      profile={
        Object {
          "bio": "test user bio",
          "location": "test user location",
          "user": "test user id",
        }
      }
    />
    <ProfileItem
      profile={
        Object {
          "bio": "test user2 bio",
          "location": "test user2 location",
          "user": "test user2 id",
        }
      }
    />
  </div>
</Fragment>
`;
